### PR TITLE
Zoom out on extruded object

### PR DESCRIPTION
**Pull request by lf94**
Closes https://github.com/KittyCAD/modeling-app/issues/2816

Video of before and after:

https://github.com/KittyCAD/modeling-app/assets/547923/5563c0d5-cfbb-4cc7-a2c4-a6068413b538

### CODE REVIEW

File: src/lang/std/engineConnection.ts
Comment (Irev-Dev) [inline src/lang/std/engineConnection.ts:2112]: There's very similar logic in `selections.ts`
```typescript
function codeToIdSelections(
  codeBasedSelections: Selection[]
): SelectionToEngine[] {
  return codeBasedSelections
    .flatMap(({ type, range, ...rest }): null | SelectionToEngine[] => {
      // TODO #868: loops over all artifacts will become inefficient at a large scale
      const entriesWithOverlap = Object.entries(
        engineCommandManager.artifactMap || {}
      )
        .map(([id, artifact]) => {
          return artifact.range && isOverlap(artifact.range, range)
            ? {
                artifact,
                selection: { type, range, ...rest },
                id,
              }
            : false
        })
        .filter(Boolean)
      let bestCandidate
      // ...
```
We might consider sharing logic between the two?

**Review by Irev-Dev — APPROVED** [review]
if we starting to  zoom to fit at the times when the user is not explicitly asking for it (right click on the gizmo, reset view, is the user asking for it, and zoom to fit on load is a special case). Then we need an animated bool on the zoom to fit endpoint.

I don't think it should block this PR, but @lf94 Would you mind opening an engine issue asking if we can animate it?

It's just a bit jarring having the camera snap vs moving into position.

**Comment by Irev-Dev** [comment]
We could test this by adding a test to `e2e/playwright/snapshot-tests.spec.ts` and taking a screen shot after an extrude is done, if we ever regress this zoom to fit, that screen shot would change.

File: src/lang/std/engineConnection.ts
Comment (Irev-Dev) [reply src/lang/std/engineConnection.ts:2112]: maybe not, they're not 1:1, I'll leave it up to you.

Comment (lf94) [reply src/lang/std/engineConnection.ts:2112]: We need to eventually rework how we get ids.

You're right these are similar but the issue with codeToIdSelections is sometimes there is code ranges that refers to two different objects. For mapRangeToObjectId, this can't happen because of the targetCommandType.

I would rather leave it and rework our code to remove both of these routines in the future.

**Comment by lf94** [comment]
> We could test this by adding a test to e2e/playwright/snapshot-tests.spec.ts and taking a screen shot after an extrude is done, if we ever regress this zoom to fit, that screen shot would change.

I've found zoom-to-fit, while deterministic, can completely change based on any angle you're viewing an object. When factoring in padding, I think this could break at any time. I'm not a fan of a test that can't guarantee robustness but I can try!

**Comment by Irev-Dev** [comment]
> I've found zoom-to-fit, while deterministic, can completely change based on any angle you're viewing an object. When factoring in padding, I think this could break at any time. I'm not a fan of a test that can't guarantee robustness but I can try!

Right I've noticed zoom to fit sometimes doing wacky things but hard to replicate.
I'm channelling my inner Jess and think we think we should try "notfeatures without tests", I'd rather us add a test and then remove it because the value-to-flakeyness ratio is too outa wack, than not even try, because what if is stable but we never find out because we didn't try.

But you have my blessing to give up at the slightest sign that the test is not going to behave itself.

File: src/lang/std/engineConnection.ts
Comment (Irev-Dev) [reply src/lang/std/engineConnection.ts:2112]: Yeah cool, relevant issue for artifactMap btw
https://github.com/KittyCAD/modeling-app/issues/868

**Comment by jessfraz** [comment]
I talked to serena and if we dont want it to keep the perspective we should use isometric view, not szoom to fit, zoom to fit will keep the cameras angle, anyways hope that helps

**Comment by jessfraz** [comment]
its "zoom to fit but at whatever andle you were at, it does not move the camera

**Comment by jessfraz** [comment]
also another fun fact if you call isometricView it shhould also do zoom to fit so you dont need two calls

**Comment by Irev-Dev** [comment]
> its "zoom to fit but at whatever angle you were at, it does not move the camera

Oh I had assumed it would translate the camera, but not change orientation (like it would have the exact same quanternion), maybe that's what you mean, it must be moving the camera.

**Comment by Irev-Dev** [comment]
but we can't change the camera to isometricView when the user extrudes something though right? that's a pretty wacky thing to do, no?

**Comment by jessfraz** [comment]
Yeah that would be pretty weird but Serena can make it work it how ever you want

**Comment by lf94** [comment]
Will add test in the morning :+1:

**Comment by lf94** [comment]
I really tried @Irev-Dev but playwright **will not let me select the drawing plane again**. I've wasted way too much time on this. Here is the code if someone wants to try:

```
test('Zoom to fit extruded object only', async ({ page, context }) => {
  // Setup utilities
  const u = await getUtils(page)
  const viewportSize = { width: 1200, height: 500 }
  const center = { x: viewportSize.width / 2, y: viewportSize.height / 2 }
  const { click00r } = getMovementUtils({ center, page })

  // Make sure axis not rendered
  await page.addInitScript(async () => {
    localStorage.setItem('disableAxis', 'true')
  })

  // Connect to engine
  await page.setViewportSize(viewportSize)
  await page.goto('/')
  await u.waitForAuthSkipAppStart()

  // Wait until sketch button is available
  let buttonSketch = page.getByRole('button', { name: 'Start Sketch' })
  await expect(buttonSketch).not.toBeDisabled()

  // Start a sketch and select plane
  await buttonSketch.click()
  await page.mouse.click(center.x, viewportSize.height * 0.55)

  // Sketch a square
  await click00r(50, 50)
  await click00r(50, 0)
  await click00r(0, 50)
  await click00r(-50, 0)
  await click00r(0, -50)
  await click00r(undefined, undefined)
  await page.getByRole('button', { name: 'Exit Sketch' }).click()

  let square1Center = { x: center.x + 50 + 25, y: center.y + 50 + 25 }

  // Wait two seconds for the UI to readjust
  await page.waitForTimeout(2 * 1000)

  // Select the square face
  await page.mouse.click(square1Center.x, square1Center.y, { delay: 200 })

  // Wait until extrude button is available
  let buttonExtrude = page.getByRole('button', { name: 'Extrude' })
  await expect(buttonExtrude).not.toBeDisabled()
  buttonExtrude.click({ delay: 200 })

  await page.waitForTimeout(2 * 1000)

  // Extrude the square by 500 units (mm)
  let cmLine = page.locator('#arg-form .cm-content')
  await expect(cmLine).toBeVisible()
  cmLine.fill('500')
  await page.locator('button[type="submit"][form="arg-form"]').click()
  // Confirm extrude
  await page.locator('button[type="submit"][form="review-form"]').click()

  // Wait two seconds for the UI to readjust
  await page.waitForTimeout(2 * 1000)

  // Repeat; should only focus on the newly extruded object

  // Wait until sketch button is available
  buttonSketch = page.getByRole('button', { name: 'Start Sketch' })
  await expect(buttonSketch).not.toBeDisabled()

  // Start a sketch and select plane
  await buttonSketch.click()

  await page.mouse.move(center.x + 5, center.y - 5, { steps: 20 })
  await page.mouse.click(center.x + 5, center.y - 5, {delay:  100 })

  // Sketch a square
  await click00r(100, 100)
  await click00r(50, 0)
  await click00r(0, 50)
  await click00r(-50, 0)
  await click00r(0, -50)
  await click00r(undefined, undefined)
  await page.getByRole('button', { name: 'Exit Sketch' }).click()

  const square2Center = { x: center.x + 100 + 25, y: center.y + 100 + 25 }

  // Wait two seconds for the UI to readjust
  await page.waitForTimeout(2 * 1000)

  // Select the square face
  await page.mouse.click(square2Center.x, square2Center.y, { delay: 200 })

  // Wait until extrude button is available
  buttonExtrude = page.getByRole('button', { name: 'Extrude' })
  await expect(buttonExtrude).not.toBeDisabled()
  buttonExtrude.click({ delay: 200 })

  // Extrude the square by 10 units (mm)
  cmLine = page.locator('#arg-form .cm-content')
  await expect(cmLine).toBeVisible()
  cmLine.fill('10')
  await page.locator('button[type="submit"][form="arg-form"]').click()
  // Confirm extrude
  await page.locator('button[type="submit"][form="review-form"]').click()

  // Wait two seconds for the UI to readjust
  await page.waitForTimeout(2 * 1000)

  // Take a screenshot
})
```

**Comment by lf94** [comment]
Can we merge

**Comment by lf94** [comment]
Ok nice, will do

**Comment by lf94** [comment]
qawolf is on creating a test for this; ill rebase in the morning :)

**Comment by lf94** [comment]
@jessfraz please force merge

**Pull request merged**
